File: ADA/syllabus/HeapSort.cpp
#include <iostream>
using namespace std;

void maxHeapify(int nums[], int n, int i) {
    int largest = i;
    int leftChild = 2 * i + 1;
    int rightChild = 2 * i + 2;

    if(leftChild < n && nums[leftChild] > nums[largest])
        largest = leftChild;
    if(rightChild < n && nums[rightChild] > nums[largest])
        largest = rightChild;
    
    if(largest != i) {
        int t = nums[largest];
        nums[largest] = nums[i];
        nums[i] = t;

        maxHeapify(nums, n, largest);
    }
}

void heapSort(int nums[], int n) {  // using MAX-Heap
    for(int i = n / 2 - 1; i >= 0; i--)
        maxHeapify(nums, n, i);

    for(int i = n - 1; i > 0; i--) {
        int t = nums[0];
        nums[0] = nums[i];
        nums[i] = t;

        maxHeapify(nums, i, 0);
    }
}

int main() {
    int nums[] = {90, 2, 1, 45, 23, 5};
    heapSort(nums, 6);
    for(int i = 0; i < 6; i++)
        cout << nums[i] << " ";
    cout << "\n";

    return 0;
}

File: Python/sorting/HeapSort.py
def maxHeapify(arr, n, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and arr[l] > arr[largest] :
        largest = l
    if r < n and arr[r] > arr[largest] :
        largest = r
    
    if largest is not i :
        arr[i], arr[largest] = arr[largest], arr[i]
        maxHeapify(arr, n, largest)

def heapSort(arr):
    n = len(arr)

    for i in range(n // 2 - 1, -1, -1):
        maxHeapify(arr, n, i)
    
    for j in range (n - 1, 0, -1):
        arr[j], arr[0] = arr[0], arr[j]
        maxHeapify(arr, j, 0)

arr = [8, 9, 1, 3, 5, 2, 4, 7, 6]
heapSort(arr)

for i in range(len(arr)):
    print(arr[i])